mlflow/evaluation.py

```diff
@@ -0,0 +1,415 @@
+            return self._hash
+        else:
+            md5_gen = hashlib.md5()
+            if isinstance(self.data, np.ndarray):
+                EvaluationDataset._gen_md5_for_arraylike_obj(md5_gen, self.data)
+                EvaluationDataset._gen_md5_for_arraylike_obj(md5_gen, self.labels)
+            elif isinstance(self.data, pd.DataFrame):
+                EvaluationDataset._gen_md5_for_arraylike_obj(md5_gen, self.data)
+                md5_gen.update(self.labels.encode("UTF-8"))
+            return md5_gen.hexdigest()
+
+    @property
+    def metadata(self):
+        metadata = {
+            "hash": self.hash,
+            "path": self.path,
+        }
+        if self.user_specified_name is not None:
+            metadata["name"] = self.user_specified_name
+        return metadata
+
+
+class GetOrCreateRunId:
+    """
+    Get or create a run, return a run_id
+    if user specified a run_id, use it.
+    otherwise if there's an active run, use it
+    otherwise create a managed run.
+    """
+
+    def __init__(self, run_id):
+        self.managed_run_context = None
+        if run_id is not None:
+            self.run_id = run_id
+        elif mlflow.active_run() is not None:
+            self.run_id = mlflow.active_run().info.run_id
+        else:
+            self.run_id = None
+
+    def __enter__(self):
+        if self.run_id is not None:
+            return self.run_id
+        else:
+            self.managed_run_context = mlflow.start_run()
+            return self.managed_run_context.__enter__().info.run_id
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        if self.managed_run_context is not None:
+            return self.managed_run_context.__exit__(exc_type, exc_val, exc_tb)
+
+
+class ModelEvaluator:
+    def can_evaluate(self, model_type, evaluator_config=None, **kwargs) -> bool:
+        """
+        :param model_type: A string describing the model type (e.g., "regressor",
+                           "classifier", …).
+        :param evaluator_config: A dictionary of additional configurations for
+                                 the evaluator.
+        :param **kwargs: For forwards compatibility, a placeholder for additional
+                         arguments that may be added to the evaluation interface
+                         in the future.
+        :return: True if the evaluator can evaluate the specified model on the
+                 specified dataset. False otherwise.
+        """
+        raise NotImplementedError()
+
+    def compute_metrics_and_compute_and_log_artifacts(
+        self, model, model_type, dataset, evaluator_config, run_id
+    ):
+        """
+        return an tuple of:
+         - an instance of EvaluationMetrics
+         - a dict of artifact_name -> instance_of_EvaluationArtifact
+        and log artifacts into run specified by run_id
+        """
+        raise NotImplementedError()
+
+    def evaluate(
+            self,
+            model: PyFuncModel,
```

Comment:
I propose to change `predict` argument to be `model` argument and force it to be `PyFuncModel` type, reason:
* force to containerize the model
* Currently `PyFuncModel` only support `predict`, if user want to support `predict_proba`, he can define custom PyFuncModel which contains `predict_proba` methods, and in concrete Evaluator implementation, we will check `hasattr(model, 'predict_proba')` to decide whether to output probability related metrics and artifacts.

And another proposal:
Shall we make `model_type` to be an attribute of `PyFuncModel` ?

Reply:
@WeichenXu123 Changing `predict()` to `model` with type `PyFuncModel` sounds good to me! 

For model_type, how would we find the right type to add as an attribute to `PyFuncModel`? Users would need some place to specify this when saving the model, right?